EFCache.Redis/RedisCache.cs

```diff
@@ -109,137 +246,211 @@ public bool GetItem(string key, out object value)
+		/// </summary>
+		/// <param name="key"></param>
 		public void InvalidateItem(string key)
 		{
 			key.GuardAgainstNullOrEmpty(nameof(key));
 
 			var database = Connection.GetDatabase();
 			var hashedKey = HashKey(key);
 
-			try
+			var entry = database.ObjectGet<CacheEntry>(key);
+			if (entry == null) return;
+			var keys = new List<RedisKey> {key};
+			keys.AddRange(entry.EntitySets.Select(AddCacheQualifier));
+
+			//TODO: Remove this!!
+			var random = new Random();
+
+			PerformWithRetryBackoff(() =>
 			{
-				var entry = database.ObjectGet<CacheEntry>(hashedKey);
-				if (entry == null) return;
+				//TODO: Remove this!!
+				// (new[] { 0 }.Contains(random.Next() % 3)) throw new Exception("Chaos monkey");
 
-				var transaction = database.CreateTransaction();
-				transaction.KeyDeleteAsync(hashedKey);
-				foreach (var set in entry.EntitySets)
-				{
-					transaction.SetRemoveAsync(AddCacheQualifier(set), hashedKey);
-				}
-				var committed = transaction.Execute();
+				database.ScriptEvaluate(_invalidateSingleyKeyScriptSha1, keys.ToArray());
+			});
 
-				if (!ShouldCollectStatistics || !committed) return;
-				
-				database.HashIncrement(AddStatsQualifier(hashedKey), InvalidationsIdentifier, 1, CommandFlags.FireAndForget);
+			// Because we don't want to include stats tracking in the lua script, we fire this off regardless of whether the invalidation
+			// succeeded. The stats may not be 100% accurate but that's the tradeoff for simplifying the primary logic
+			if (!ShouldCollectStatistics) return;
+
+			database.HashIncrement(AddStatsQualifier(hashedKey), InvalidationsIdentifier, 1, CommandFlags.FireAndForget);
+		}
+
+		#endregion
+
+		#region ILockableCache
+
+		/// <summary>
+		/// Use the RedLock algorithm to create a mutex lock on all of the affected entity sets with
+		/// generally sane default expiry, wait, and retry values. If locks are not acquired on all
+		/// entity sets, return null indicating lock was not possible. Deadlocks are actually possible
+		/// here. With reasonable wait and retry values, deadlocks should not result in failure.
+		///
+		/// If consistency is a requirement in your application, use this method. These locks should
+		/// be acquired prior to any database insert, update, or delete. Once the
+		/// db operation is successful, the locks should be released.
+		/// </summary>
+		/// <param name="entitySets">Entity sets to acquire locks for</param>
+		/// <returns>A set of IRedLock objects</returns>
+		public object Lock(IEnumerable<string> entitySets)
+		{
+			// Lazy eval the database in case the connection hasn't been instantiated yet
+			Connection.GetDatabase();
+
+			var expiry = TimeSpan.FromSeconds(30);
+			var wait = TimeSpan.FromSeconds(10);
+			var retry = TimeSpan.FromSeconds(1);
+
+			var redLocks = new List<IRedLock>();
+			redLocks.AddRange(entitySets.Select(entitySet => LazyRedLockFactory.Value.CreateLock(entitySet, expiry, wait, retry)));
```

Comment:
@jkdufair is the first param of `CreateLock` supposed to be the key of the set? If so I think we need to use `AddCacheQualifier`

Reply:
In this case, no. It just needs to be something unique to identify the entity set. The RedLock algorithm just creates a random key in Redis and associates it with a local object. Basically just a token we use to keep track of what's locked and what's not.